expertise/models/specter2_scincl/scincl.py

```diff
@@ -0,0 +1,347 @@
+            paper_num_test = len(test_id_list)
+
+        print('Computing all scores...')
+        p2p_aff = torch.empty((paper_num_test, paper_num_train), device=torch.device('cpu'))
+        for i in range(paper_num_test):
+            p2p_aff[i, :] = torch.sum(paper_emb_test[i, :].unsqueeze(dim=0) * paper_emb_train, dim=1)
+
+        if self.dump_p2p:
+            p2p_dict = {}
+            for i in range(paper_num_test):
+                p2p_dict[test_id_list[i]] = {}
+                for j in range(paper_num_train):
+                    p2p_dict[test_id_list[i]][train_id_list[j]] = float(p2p_aff[i, j])
+            with open(p2p_path, 'w') as f:
+                json.dump(p2p_dict, f, indent=4)
+
+        # Normalize all scores
+        min_val = p2p_aff.min()
+        max_val = p2p_aff.max()
+        p2p_aff_norm = (p2p_aff - min_val) / (max_val - min_val)
+
+        csv_scores = []
+        self.preliminary_scores = []
+        for reviewer_id, train_note_id_list in self.pub_author_ids_to_note_id.items():
+            if len(train_note_id_list) == 0:
+                continue
+            train_paper_idx = []
+            for paper_id in train_note_id_list:
+                if paper_id not in train_bad_id_set:
+                    train_paper_idx.append(paper_id2train_idx[paper_id])
+            train_paper_aff_j = p2p_aff_norm[:, train_paper_idx]
+
+            if self.average_score:
+                all_paper_aff = train_paper_aff_j.mean(dim=1)
+            elif self.max_score:
+                all_paper_aff = train_paper_aff_j.max(dim=1)[0]
+            for j in range(paper_num_test):
+                csv_line = '{note_id},{reviewer},{score}'.format(note_id=test_id_list[j], reviewer=reviewer_id,
+                                                                 score=all_paper_aff[j].item())
+                csv_scores.append(csv_line)
+                self.preliminary_scores.append((test_id_list[j], reviewer_id, all_paper_aff[j].item()))
+
+        if scores_path:
+            with open(scores_path, 'w') as f:
+                for csv_line in csv_scores:
+                    f.write(csv_line + '\n')
+
+        return self.preliminary_scores
+
+    def _sparse_scores_helper(self, all_scores, id_index):
+        counter = 0
+        # Get the first note_id or profile_id
+        current_id = self.preliminary_scores[0][id_index]
+        if id_index == 0:
+            desc = 'Note IDs'
+        else:
+            desc = 'Profiles IDs'
+        for note_id, profile_id, score in tqdm(self.preliminary_scores, total=len(self.preliminary_scores), desc=desc):
+            if counter < self.sparse_value:
+                all_scores.add((note_id, profile_id, score))
+            elif (note_id, profile_id)[id_index] != current_id:
+                counter = 0
+                all_scores.add((note_id, profile_id, score))
+                current_id = (note_id, profile_id)[id_index]
+            counter += 1
+        return all_scores
```

Comment:
Is this method being repeated in every model class? Maybe we should start extracting this methods and place them in a common file, unless there is a good reason not to do it.